### ISSUE TITLE
Multiple `var` for the same binding should not be added in one function/module scope

### ISSUE THREAD

**Issue by eventualbuddha** [eventualbuddha/add-variable-declarations#8]
From https://github.com/decaffeinate/decaffeinate/issues/316:

``` js
// This input…
let fn = function(a, b) {
  if (a) {
    if (b) {
      r = 0;
    } else {
      r = 1;
    }
  } else {
    r = 2;
  }
  return r;
};

// …yields this output:
let fn = function(a, b) {
  if (a) {
    if (b) {
      var r = 0;
    } else {
      var r = 1;
    }
  } else {
    var r = 2;
  }
  return r;
};
```

There is a failing test for this [here](https://github.com/eventualbuddha/add-variable-declarations/blob/master/test/fixtures/multiple-assignments-in-nested-scopes-hoists-to-shared-scope/input.js).

As @avonwyss pointed out, this is bad style and is flagged as such by static analysis tools. We should ensure that we add `var` for a particular binding only once per function or module and that it is in the innermost scope that it can be. That is, the example above should be this:

``` js
let fn = function(a, b) {
  var r;
  if (a) {
    if (b) {
      r = 0;
    } else {
      r = 1;
    }
  } else {
    r = 2;
  }
  return r;
};
```

But here the `var` should be in a nested scope:

``` js
// input
if (a) {
  if (b) {
    r = 0;
  } else {
    r = 1;
  }
}

// output
if (a) {
  var r;
  if (b) {
    r = 0;
  } else {
    r = 1;
  }
}
```

**Comment by alangpierce** [comment]
I'm working on this (and also https://github.com/decaffeinate/decaffeinate/issues/605 ), and it turns out to be a pretty significant change to this project, so I guess I'll write down some notes of my plan in case I'm overlooking a better way:

* Rather than ever doing magic-string operations while traversing the tree, the new plan is to maintain tentative information about all bindings, which might change as we discover more information, and then only "commit" those bindings at the end (or, at least, when we know they won't change anymore).
* Each of these tentative binding objects is "owned" by some function (or program) scope (really some TraverseState with a function scope), even if we end up adding the `var` in a more specific scope. That means we won't miss declarations within `if` statements (fixing https://github.com/decaffeinate/decaffeinate/issues/605 ) and lets us know that two variables are the same for the purpose of finding the most specific scope.
* As we see variable declarations and usages, we traverse up to the enclosing function scope to find the latest binding information, then update the "most specific scope" information to the lowest common ancestor of the two.
* In cases where we'd prefer to do an inline declaration (e.g. a destructure), we only end up doing so if none of the assigned variables ended up moving to a parent scope.
